fxfs/core/api/privatefs.go

```diff
@@ -0,0 +1,214 @@
+package api
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	pfs "github.com/functionland/go-fula/fxfs/core/pfs"
+	"github.com/ipfs/go-blockservice"
+	"github.com/ipfs/go-cidutil"
+	"github.com/ipfs/go-filestore"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+	files "github.com/ipfs/go-ipfs-files"
+	ipld "github.com/ipfs/go-ipld-format"
+	dag "github.com/ipfs/go-merkledag"
+	dagtest "github.com/ipfs/go-merkledag/test"
+	"github.com/ipfs/go-mfs"
+	ft "github.com/ipfs/go-unixfs"
+	unixfile "github.com/ipfs/go-unixfs/file"
+	"github.com/ipfs/interface-go-ipfs-core/options"
+	"github.com/ipfs/interface-go-ipfs-core/path"
+	"github.com/ipfs/kubo/core"
+	"github.com/ipfs/kubo/tracing"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+)
+
+type PrivateAPI CoreAPI
+
+func (api *PrivateAPI) Add(ctx context.Context, files files.Node, opts ...options.UnixfsAddOption) (path.Resolved, error) {
+	ctx, span := tracing.Span(ctx, "CoreAPI.UnixfsAPI", "Add")
+	defer span.End()
+
+	settings, prefix, err := options.UnixfsAddOptions(opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	span.SetAttributes(
+		attribute.String("chunker", settings.Chunker),
+		attribute.Int("cidversion", settings.CidVersion),
+		attribute.Bool("inline", settings.Inline),
+		attribute.Int("inlinelimit", settings.InlineLimit),
+		attribute.Bool("rawleaves", settings.RawLeaves),
+		attribute.Bool("rawleavesset", settings.RawLeavesSet),
+		attribute.Int("layout", int(settings.Layout)),
+		attribute.Bool("pin", settings.Pin),
+		attribute.Bool("onlyhash", settings.OnlyHash),
+		attribute.Bool("fscache", settings.FsCache),
+		attribute.Bool("nocopy", settings.NoCopy),
+		attribute.Bool("silent", settings.Silent),
+		attribute.Bool("progress", settings.Progress),
+	)
+
+	cfg, err := api.repo.Config()
+	if err != nil {
+		return nil, err
+	}
+
+	// check if repo will exceed storage limit if added
+	// TODO: this doesn't handle the case if the hashed file is already in blocks (deduplicated)
+	// TODO: conditional GC is disabled due to it is somehow not possible to pass the size to the daemon
+	//if err := corerepo.ConditionalGC(req.Context(), n, uint64(size)); err != nil {
+	//	res.SetError(err, cmds.ErrNormal)
+	//	return
+	//}
+
+	if settings.NoCopy && !(cfg.Experimental.FilestoreEnabled || cfg.Experimental.UrlstoreEnabled) {
+		return nil, fmt.Errorf("either the filestore or the urlstore must be enabled to use nocopy, see: https://github.com/ipfs/kubo/blob/master/docs/experimental-features.md#ipfs-filestore")
```

Comment:
Use `errors.New` when there is nothing to format. There are quite a few cases like this across the PR. Have a search in other places.

Reply:
The original code in IPFS coreapi uses `fmt.Errorf`. But it seems more correct this way (using `errors.New` since the signature explicitly declares that output is of type error.
Done